aip/client-libraries/4222.md

```diff
@@ -33,14 +33,21 @@ rpc CreateTopic(CreateTopicRequest) {
 In this case, the applicable variable is `parent`, and it refers to the
 `parent` field in `CreateTopicRequest`. When the user provides an instance of
 `CreateTopicRequest` to the method (or once the client library has built it, in
-the case of method overloads), the client library should then extract the key
-and value, URL-encode them, and append them to the `x-goog-request-params`
-header. Much like URL parameters, if there is more than one key-value pair, the
-`&` character is used as the separator.
+the case of method overloads), the client library must extract the key
+and value, and append them to the `x-goog-request-params` header. The key-value
```

Comment:
Do we encode both keys and values, or just values?

Reply:
**tl;dr**: we should encode both keys and values (the entire query component)

The [RFC 6570 §3.2.2](https://tools.ietf.org/html/rfc6570#section-3.2.2) Hong mentioned focuses on variable expansion as replacement of a variable name (key) with a value. But really, here we are building a query string.

[RFC 3986](https://tools.ietf.org/html/rfc3986) defining URIs implies entire components (e.g. a "query" => `?field=value`) should be encoded.

In the context of `x-www-form-urlencoded`, [whatwg's standard](https://url.spec.whatwg.org/#urlencoded-serializing) states:
> Let `name` be the result of serializing the result of encoding `tuple`’s `name`, using encoding

(where `encoding` is [percent-encoding](https://url.spec.whatwg.org/#percent-encode))

In the context of `multipart/form-data` [RFC 1867 §5.11](https://www.ietf.org/rfc/rfc1867.txt) explicitly states non-ASCII `field names should be encoded`.